src/core/vector/qgsvectorlayerfeatureiterator.cpp

```diff
@@ -114,6 +114,7 @@ QgsVectorLayerFeatureSource::~QgsVectorLayerFeatureSource() = default;
 
 QgsFeatureIterator QgsVectorLayerFeatureSource::getFeatures( const QgsFeatureRequest &request )
 {
+  QMutexLocker mutexLocker( &mGetFeatureMutex );
```

Comment:
I don't think this is the correct place for the mutex -- if you look at QgsAbstractFeatureIteratorFromSource it's calling methods on the source, so each iterator could potentially be modifying the source even after they were created in a thread-safe manner. 

I was thinking more that the thread safety logic should live in the expression/expression utils. Eg remove the shared pointer, and (with the mutex protection) only let the first call (or first thread containing callers?) return the feature source previously stored. And then other callers get the existing fallback logic...

Reply:
Just briefly looking through the code of the feature sources, it seems like almost all access is read only and thread safe. It could be an alternative to just make `mSource` `const` for iterators.
I would assume that some providers write back some data (e.g. lookuptables for feature ids, local caches), but these are probably already now mostly shared in a thread safe way.

Intuitively, this sounds more performant and cleaner.
Do you see any blockers to this approach?